editor/plugins/node_3d_editor_gizmos.cpp

```diff
@@ -2121,52 +2121,60 @@ void Skeleton3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
 	Color bonecolor = Color(1.0, 0.4, 0.4, 0.3);
 	Color rootcolor = Color(0.4, 1.0, 0.4, 0.1);
 
-	for (int i_bone = 0; i_bone < skel->get_bone_count(); i_bone++) {
-		int i = skel->get_process_order(i_bone);
+	Vector<int> bones_to_process = skel->get_parentless_bones();
+	while (bones_to_process.size() > 0) {
+		int current_bone_idx = bones_to_process[0];
+		bones_to_process.erase(current_bone_idx);
 
-		int parent = skel->get_bone_parent(i);
+		Vector<int> child_bones_vector = skel->get_bone_children(current_bone_idx);
+		int child_bones_size = child_bones_vector.size();
 
-		if (parent >= 0) {
-			grests.write[i] = grests[parent] * skel->get_bone_rest(i);
+		// You have children but no parent, then you must be a root/parentless bone.
+		if (child_bones_size >= 0 && skel->get_bone_parent(current_bone_idx) <= 0) {
+			grests.write[current_bone_idx] = skel->global_pose_to_local_pose(current_bone_idx, skel->get_bone_global_pose(current_bone_idx));
+		}
+
+		for (int i = 0; i < child_bones_size; i++) {
+			int child_bone_idx = child_bones_vector[i];
 
-			Vector3 v0 = grests[parent].origin;
-			Vector3 v1 = grests[i].origin;
-			Vector3 d = (v1 - v0).normalized();
-			float dist = v0.distance_to(v1);
+			grests.write[child_bone_idx] = skel->global_pose_to_local_pose(child_bone_idx, skel->get_bone_global_pose(child_bone_idx));
+			Vector3 v0 = grests[current_bone_idx].origin;
+			Vector3 v1 = grests[child_bone_idx].origin;
+			Vector3 d = skel->get_bone_rest(child_bone_idx).origin.normalized();
+			float dist = skel->get_bone_rest(child_bone_idx).origin.length();
 
-			//find closest axis
+			// Find closest axis.
 			int closest = -1;
 			float closest_d = 0.0;
-
 			for (int j = 0; j < 3; j++) {
-				float dp = Math::abs(grests[parent].basis[j].normalized().dot(d));
+				float dp = Math::abs(grests[current_bone_idx].basis[j].normalized().dot(d));
 				if (j == 0 || dp > closest_d) {
 					closest = j;
 				}
 			}
 
-			//find closest other
```

Comment:
Placeholder for all the lines to have full sentence case and end with punctuation.

Reply:
See https://docs.godotengine.org/en/stable/community/contributing/code_style_guidelines.html#comment-style-guide.

Reply:
I added a period to the end, but I cannot really expand or elaborate on the comment, as it's from code I didn't write so I'm not  100% positive what the comment points to.